[PLATFORM-475] Share session across tabs, redirect to login on auth failure

timoxley opened the pull request:
Converts session provider to use `window.localStorage` instead of `window.sessionStorage`. This fixes the core issue of needing to login when you open multiple tabs.
To handle the logout case I've added a redirect to the login page (and back to current page) on detecting a 401 response from any API endpoint as discussed.

There may be issues with this change should any API responses respond with a 401 and a redirect to login is not appropriate. I couldn't see any references to a 401 status checks in the existing code, but it's possible that there's some non-discriminating error handler that will no longer be run due to this interceptor. Perhaps not.

timoxley commented:
> Requesting changes because the root page and the marketplace should be usable without logging in.

Ah of course. 👍

timoxley commented:
Ahhh, I wanted to determine whether a redirect was necessary based on the path, e.g. if in /core/*, or /canvas/* or on a product edit route, then do the redirect, which will mostly work, but I'm not sure how to do the check appropriately for things like public canvases. Blegh.

timoxley commented:
now I think about it, I hit this exact same issue last time I tried to fix this

juhah commented:
Yeah, it needs to be smart about the routes... Perhaps some wrapper component for the protected routes that does the redirect / pops up the login screen.

mondoreale commented on app/src/auth/components/SessionProvider/index.jsx:
```diff
@@ -2,60 +2,63 @@
+    if (value) {
+        storage.setItem(SESSION_TOKEN_KEY, value)
+    } else {
+        // remove entire key if not set
```
👍

mondoreale commented on app/src/auth/components/SessionProvider/index.jsx:
```diff
@@ -2,60 +2,63 @@
-        }
-        return null
-    }
+let cachedToken // fallback if no webstorage
```
Ha! Good call.

timoxley commented on app/src/setup.js:
```diff
@@ -2,6 +2,23 @@
+        } : {})
+    }
+    // otherwise ignore
+    throw error
```
Good point.

timoxley replied:
update: it should probably still throw so the receiver sees a failure rather than success

timoxley commented:
@juhah ok I've rebased and updated the logic please have good try at breaking this again. It now redirects to login on failed non-users/me api call. Logging out from a different tab should fail on first failing api response.

timoxley commented:
@fonty1 added you as a reviewer, please try break this.

juhah commented on app/src/auth/utils/loginInterceptor.js:
```diff
@@ -0,0 +1,48 @@
+import axios from 'axios'
+import routes from '$routes'
+
+const meURL = `${process.env.STREAMR_API_URL}/users/me`
```
Let's wrap API urls with `formatApiUrl('users', 'me')`

timoxley replied:
Done in eea635e

fonty1 commented:
> @fonty1 added you as a reviewer, please try break this.

I could break it like Juha suggested, but not terribly fussed about that corner case either. The notification about being 'not auth'd by token or **cookie**' bothered me since we're not using cookies anymore.

fonty1 commented:
I suppose I can also be auth'd on two tabs, log out of one, then try logout of the other, which will crash the app. After the crash the tab will still think I'm logged in which is a little weird.

timoxley commented:
@fonty1 @juhah 

* "Fixed" issue with the crashing when logging out again by delaying the error for a moment, which should allow the redirect to kick in.
* Set it to redirect to / rather than login if the failure occurs when the user is on the logout route. 
* Prevented it from immediately redirecting to logout after logging in.

timoxley commented:
> The notification about being 'not auth'd by token or cookie' bothered me since we're not using cookies anymore.

Yeah the notification content comes from the server so 🤷‍♀

timoxley commented:
> It's not a huge issue and I'm fine with leaving it like this.

let's do that for now
